crates/mun_language_server/src/main_loop.rs

```diff
@@ -1,26 +1,220 @@
-use crate::protocol::{Connection, Message};
+use crate::analysis::{Analysis, AnalysisSnapshot};
+use crate::change::AnalysisChange;
+use crate::config::{Config, FilesWatcher};
+use crate::conversion::{convert_range, url_from_path_with_drive_lowercasing};
+use crate::protocol::{Connection, Message, Notification, Request, RequestId};
 use crate::Result;
-use futures::StreamExt;
+use anyhow::anyhow;
+use async_std::sync::RwLock;
+use futures::channel::mpsc::{unbounded, Sender, UnboundedReceiver, UnboundedSender};
+use futures::{SinkExt, StreamExt};
+use lsp_types::notification::PublishDiagnostics;
+use lsp_types::{PublishDiagnosticsParams, Url};
+use ra_vfs::{RootEntry, Vfs, VfsChange, VfsFile};
+use serde::{de::DeserializeOwned, Serialize};
+use std::collections::HashSet;
+use std::sync::Arc;
 
+/// A `Task` is something that is send from async tasks to the entry point for processing. This
+/// enables synchronizing resources like the connection with the client.
+#[derive(Debug)]
+enum Task {
+    Notify(Notification),
+}
+
+#[derive(Debug)]
 enum Event {
     Msg(Message),
+    Vfs(ra_vfs::VfsTask),
+    Task(Task),
+}
+
+/// State for the language server
+struct LanguageServerState {
+    /// Interface to the vfs, a virtual filesystem that supports the overlaying of files
+    pub vfs: Arc<RwLock<Vfs>>,
+
+    /// Receiver channel to apply filesystem changes on `vfs`
+    pub vfs_task_receiver: UnboundedReceiver<ra_vfs::VfsTask>,
+
+    /// Holds the state of the analysis process
+    pub analysis: Analysis,
+
+    /// All the roots in the workspace
+    pub local_source_roots: Vec<hir::SourceRootId>,
+}
+
+/// A snapshot of the state of the language server
+struct LanguageServerSnapshot {
+    /// Interface to the vfs, a virtual filesystem that supports the overlaying of files
+    pub vfs: Arc<RwLock<Vfs>>,
+
+    /// Holds the state of the analysis process
+    pub analysis: AnalysisSnapshot,
+
+    /// All the roots in the workspace
+    pub local_source_roots: Vec<hir::SourceRootId>,
+}
+
+/// State maintained for the connection. This includes everything that is required to be able to
+/// properly communicate with the client but has nothing to do with any Mun related state.
+struct ConnectionState {
+    connection: Connection,
+
+    next_request_id: u64,
+    pending_responses: HashSet<RequestId>,
+}
+
+impl ConnectionState {
+    /// Constructs a new `ConnectionState`
+    fn new(connection: Connection) -> Self {
+        Self {
+            connection,
+            next_request_id: 0,
+            pending_responses: Default::default(),
+        }
+    }
+
+    /// Constructs a new request ID and stores that we are still awaiting a response.
+    fn next_request_id(&mut self) -> RequestId {
+        self.next_request_id += 1;
+        let res: RequestId = self.next_request_id.into();
+        let inserted = self.pending_responses.insert(res.clone());
+        assert!(inserted);
```

Comment:
```suggestion
        debug_assert!(inserted);
```

Reply:
It seems like this shouldn't actually fail at runtime. The `debug_assert!` macro would only include this in debug builds, saving some performance in release builds.